consumer/file.go

```diff
@@ -140,351 +139,118 @@ func init() {
-
-	if cons.source.file != nil {
-		cons.source.file.Close()
-		cons.source.file = nil
-		cons.seeker.seek = cons.seeker.onRotate
-		cons.seeker.offset = 0
-		if cons.source.offsetFileName != "" {
-			cons.storeOffset()
-		}
-	}
+func (cons *File) newObservedFile(name string) *observableFile {
+	logger := cons.Logger.WithFields(logrus.Fields{
+		"File": name,
+	})
 
-	if cons.source.offsetFileName != "" {
-		fileContents, err := ioutil.ReadFile(cons.source.offsetFileName)
-		if err == nil {
-			cons.seeker.seek = 1
-			cons.seeker.offset, err = strconv.ParseInt(string(fileContents), 10, 64)
-			if err != nil {
-				cons.Logger.Error("Error reading offset file: ", err)
+	offsetFileName := ""
+	defaultOffset := strings.ToLower(cons.defaultOffset)
+	cursor := fileCursor{whence: io.SeekStart}
+
+	switch {
+	case cons.offsetFilePath != "":
+		offsetFileName = fmt.Sprintf("%s/%s.offset", cons.offsetFilePath, filepath.Base(cons.fileName))
+		if offsetFileData, err := ioutil.ReadFile(offsetFileName); err == nil {
+			logger.WithError(err).Errorf("Failed to open offset file %s", offsetFileName)
+		} else {
+			if offset, err := strconv.ParseInt(string(offsetFileData), 10, 64); err != nil {
+				logger.WithError(err).Errorf("Error reading offset number from %s", offsetFileName)
+			} else {
+				cursor.offset = offset
 			}
 		}
-	}
-}
-
-func (cons *File) close() {
-	if cons.source.file != nil {
-		cons.source.file.Close()
-	}
-	cons.setState(fileStateDone)
-	cons.WorkerDone()
-}
 
-func (cons *File) observe() {
-	defer cons.close()
-
-	sendFunction := cons.Enqueue
-	if cons.source.offsetFileName != "" {
-		sendFunction = cons.enqueueAndPersist
+	case defaultOffset == fileOffsetEnd:
+		cursor.whence = io.SeekEnd
 	}
 
-	buffer := tio.NewBufferedReader(fileBufferGrowSize, 0, 0, cons.delimiter)
+	logger.Info("Starting file scraper")
 
-	cons.Logger.WithField("file", cons.source.realFileName).Debugf("Use observe mode '%s'", cons.observeMode)
-	if cons.observeMode == observeModeWatch {
-		cons.watcher = newWatcher(cons.Logger, &cons.source, func() { cons.read(buffer, sendFunction, func() {}, func() {}) })
-		cons.watcher.Watch(buffer, sendFunction)
-	} else {
-		cons.poll(buffer, sendFunction)
+	return &observableFile{
+		fileName:       name,
+		offsetFileName: offsetFileName,
+		cursor:         cursor,
+		retryDelay:     cons.retryDelay,
+		pollDelay:      cons.pollingDelay,
+		buffer:         tio.NewBufferedReader(fileBufferGrowSize, tio.BufferedReaderFlagDelimiter, 0, cons.delimiter),
+		log:            logger,
 	}
 }
 
-func (cons *File) poll(buffer *tio.BufferedReader, sendFunction func(data []byte)) {
-	spin := tsync.NewCustomSpinner(cons.source.pollingDelay)
+func (cons *File) observeFile(name string) {
+	defer cons.WorkerDone()
+	file := cons.newObservedFile(name)
+	defer file.close()
 
-	for cons.source.state != fileStateDone {
-		cons.read(buffer, sendFunction, spin.Yield, spin.Reset)
-	}
-}
+	enqueue := cons.SimpleConsumer.Enqueue
```

Comment:
I do not see an obvious reason to use `cons.SimpleConsumer.Enqueue` here instead of `cons.Enqueue`, when you are using `const.EnqueueWithMetadata` a few lines below. It would be great if you could decide on one style of access for both.